File: LongestSubstringWithoutDuplicates/longest_substring_without_duplicates.py
#!/usr/bin/env python3

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        head = 0
        max_len = 0
        substr_chars = set()

        for i in range(len(s)):
            while s[i] in substr_chars:
                substr_chars.remove(s[head])
                head += 1
            substr_chars.add(s[i])
            max_len = max(len(substr_chars), max_len)

        return max_len

File: LongestSubstringWithoutDuplicates/longest_substring_without_duplicates.cpp
#include <unordered_set>
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        std::unordered_set<char> seen;
        int l = 0, max_len = 0;
        for (int r = 0; r < s.size(); ++r){
            if (seen.find(s[r]) != seen.end()){
                while (s[l] != s[r]){
                    seen.erase(s[l]);
                    ++l;
                }
                ++l;
            }
            else{
                seen.insert(s[r]);
                max_len = std::max(r - l + 1, max_len);
            }
        }
        return max_len;
    }
};

